### PR TITLE
io32_memcpy

### PR DESCRIPTION
**Pull request by SebastianBoe**
Implement memcpy32, a version of memcpy that only performs 32 bit
word-wise copying. (useful for MMIO).

Based on Pete Dietl's implementation and the mcan implementation.

Also make use of it from mcan.

I was not able to find a way to build mcan.

But I have done some simple testing locally.

A second attempt at https://github.com/zephyrproject-rtos/zephyr/pull/41273

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Are you doing this specifically to have a memcpy32 for IO purposes?

If so, should rename to memcpy_io32 or something like that.

**Review by alexanderwachter — DISMISSED** [review]
Looks good to me

**Review by andyross — COMMENTED** [review]
Nitpickery: Can we name this something less generic?  Maybe `sys_wordcopy32()` or something?  Calling it "memcpy" and putting it in util.h makes it sound like something regular apps are expected to use.  And while, sure, this can be implemented in pure C it exists only to honor MMIO requirements of hardware with behavior way outside the C standard.

Also, note that this is sort of a can of worms.  What if the hardware has ordering requirements?  Does the copy work identically forward and backwards?  Hardware can have prefetch units and write combining that messes this up even if the software is doing everything right.

I guess my worry is this: this meets the requirements of that one piece of MCAN hardware (and on the particular SOC integration!).  How many other devices are actually going to productively use it?  My guess "not enough to make this a standard-looking memcpy".

File: lib/os/Kconfig [hunk 1]
```diff
@@ -157,6 +157,14 @@ config UTF8
 	  Enable the utf8 API. The API implements functions to specifically
 	  handle UTF-8 encoded strings.
 
+config MEM32
```
Comment (stephanosio) [inline lib/os/Kconfig:160]:
```suggestion
config MEMCPY32
```

File: lib/os/Kconfig [hunk 2]
```diff
@@ -157,6 +157,14 @@ config UTF8
...
+	bool "Enable memcpy32()"
```
Comment (stephanosio) [inline lib/os/Kconfig:161]: "Enable" is redundant.


```suggestion
	bool "memcpy32"
```

**Comment by alexanderwachter** [comment]
@andyross m_can is actually a driver backend used on many SoC.
But this is just a remark.

**Comment by henrikbrixandersen** [comment]
@SebastianBoe What use-case do you have for this outside of the M_CAN driver? I tend to agree with @andyross.

File: lib/os/Kconfig (hunk 1 above)
Comment (SebastianBoe) [reply lib/os/Kconfig:160]: The idea was that compiling a C function is cheap, but invoking gcc on a source file is expensive.

So I want to gather related, future, functions in this mem32 source file.

Different alignment requirements, granularity, memset, etc.

**Comment by SebastianBoe** [comment]
@henrikbrixandersen : My use-case was to copy data from nRF->UICR->OTP MMIO, that had this limitation, to RAM. So in my case I eventuallly found that it's awkward that the function requires the buffer in RAM to be word-aligned.

I'm not sure about what to do about that. Maybe it's trivial to enforce word-aligned RAM on the buffer being written to.

Maybe the MCAN use-case also actually only needs word-aligned enforcement on the read, and not the write.

I intend to investigate and address comments.

But I want to wait a bit for feedback so I don't have to revisit this many times.

**Comment by SebastianBoe** [comment]
"Can we name this something less generic? Maybe sys_wordcopy32() or something?"

I'd like to keep the term memcpy and 32, but we can add io, or sys, or both.

memcpy32 can be familiar to other developers:
https://chromium.googlesource.com/chromiumos/platform/punybench/+/refs/heads/stabilize-gnawty-5841.84.B/cpu.m/memcpy.c
https://lwn.net/Articles/165519/

and clearly communicates that it's a variant of memcpy. wordcopy is not as clear.

**Comment by alexanderwachter** [comment]
@SebastianBoe the reason why we need to copy in 32bit chunks is that the lower two address lines are not present in the m_can. If you use the byte wise copy, all four bytes of the 32bits end up in the first byte.

**Comment by SebastianBoe** [comment]
All feedback has been addressed.

@andyross , @galak

File: lib/os/Kconfig [hunk 3]
```diff
@@ -157,6 +157,14 @@ config UTF8
 	  Enable the utf8 API. The API implements functions to specifically
 	  handle UTF-8 encoded strings.
 
+config IO32_MEM
+	bool "memcpy32()"
```
Comment (pdgendt) [inline lib/os/Kconfig:161]: The name does not match the actual function name.

**Review by pdgendt — COMMENTED** [review]
Please update the commit subjects to indicate the actual changes
- `util: memcpy32`
- `drivers: mcan`

Aren't useful

File: include/zephyr/sys/util.h [hunk 4]
```diff
@@ -498,6 +499,26 @@ char *utf8_trunc(char *utf8_str);
...
+ *
+ * NOTE: This function uses __ASSERT to enforce the requirements
+ * of each parameter given below.
+ *
+ * @param dest Destination address (must be aligned to a 4-byte boundary)
+ * @param src  Source address (must be aligned to a 4-byte boundary)
+ * @param n    number of bytes to copy (must be a multiple of 4)
+ *
+ * @return destination pointer or NULL on error
+ *
```
Comment (cfriedt) [inline include/zephyr/sys/util.h:518]: probably can remove this line

**Review by andyross — DISMISSED** [review]
I still have a vague sense that this is too specific an API to be useful, but the name is no longer ambiguous and I don't see why it shouldn't be merged.

File: lib/os/Kconfig [hunk 5]
```diff
@@ -157,6 +157,14 @@ config UTF8
 	  Enable the utf8 API. The API implements functions to specifically
 	  handle UTF-8 encoded strings.
 
+config IO32_MEM
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply lib/os/Kconfig:160]: Agreed, @SebastianBoe is this really needed?

File: lib/os/Kconfig (hunk 3 above)
Comment (carlescufi) [reply lib/os/Kconfig:161]: Yep, @SebastianBoe FYI

**Comment by alexanderwachter** [comment]
I have two comments still open. Bit no blockers.

**Review by galak — DISMISSED** [review]

File: include/zephyr/sys/util.h [hunk 6]
```diff
@@ -498,6 +499,26 @@ char *utf8_trunc(char *utf8_str);
...
+ */
+volatile void *io32_memcpy(volatile void *ZRESTRICT dest, const volatile void *ZRESTRICT src, size_t n);
```
Comment (stephanosio) [inline include/zephyr/sys/util.h:520]: What would be a reason not to introduce `io32_memset` here if we are doing this?

File: lib/os/Kconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (SebastianBoe) [reply lib/os/Kconfig:160]: Yes, there is no pre-existing source file that the function belongs in.

And we don't want to unconditionally invoke gcc for something that is rarely used as it slows down the build and creates noise in the build.

File: include/zephyr/sys/util.h (hunk 6 above)
Comment (SebastianBoe) [reply include/zephyr/sys/util.h:520]: Out-of-scope for this PR.

A PR that adds this would be welcome.

**Review by stephanosio — CHANGES_REQUESTED** [review]
There needs to be a test for this.

**Comment by SebastianBoe** [comment]
"There needs to be a test for this."
I disagree.

It is trivial enough to be proof-read to be correct. And it is system-tested as it is used by mcan.

**Comment by stephanosio** [comment]
> I disagree.
> 
> It is trivial enough to be proof-read to be correct. And it is system-tested as it is used by mcan.

While the function itself is simple enough, who is to say it will not break on some architectures and toolchains because of some ABI-specific details and/or compiler optimisations?

It is very important that we sanity-check/validate basic functions like these to ensure that they behave as intended.

Also, this is technically an API. Every API function ought to have a corresponding test.

**Comment by cfriedt** [comment]
> "There needs to be a test for this." I disagree.
> 
> It is trivial enough to be proof-read to be correct. And it is system-tested as it is used by mcan.

I was just bitten by this. Code was "trivial enough to be proof-read to be correct" but it actually introduced a subtle bug that could have been easily avoided by including a test.

Pay your future self and the future selves of maintainers please 😁

**Comment by SebastianBoe** [comment]
@stephanosio : Unfortunately I am unable to prioritize spending time adding a test for this. Please, in future reviews, try to leave all comments in one pass to avoid multiple review iterations.

**Pull request closed without merging**